extractor/filesystem/os/kernel/module/module.go

```
@@ -0,0 +1,282 @@
+		Path:          path,
+		Result:        result,
+		FileSizeBytes: fileSizeBytes,
+	})
+}
+
+// Extract extracts packages from .ko files passed through the scan input.
+func (e Extractor) Extract(ctx context.Context, input *filesystem.ScanInput) ([]*extractor.Inventory, error) {
+	inventory, err := e.extractFromInput(ctx, input)
+
+	if e.stats != nil {
+		var fileSizeBytes int64
+		if input.Info != nil {
+			fileSizeBytes = input.Info.Size()
+		}
+		e.stats.AfterFileExtracted(e.Name(), &stats.FileExtractedStats{
+			Path:          input.Path,
+			Result:        filesystem.ExtractorErrorToFileExtractedResult(err),
+			FileSizeBytes: fileSizeBytes,
+		})
+	}
+	return inventory, err
+}
+
+func (e Extractor) extractFromInput(ctx context.Context, input *filesystem.ScanInput) ([]*extractor.Inventory, error) {
+	pkgs := []*extractor.Inventory{}
+
+	m, err := osrelease.GetOSRelease(input.FS)
+	if err != nil {
+		log.Errorf("osrelease.ParseOsRelease(): %v", err)
+	}
+
+	// Return if canceled or exceeding deadline.
+	if err := ctx.Err(); err != nil {
+		return pkgs, fmt.Errorf("%s halted at %q because of context error: %v", e.Name(), input.Path, err)
+	}
+
+	data, err := io.ReadAll(input.Reader)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read input: %v", err)
+	}
+
+	elfFile, err := elf.NewFile(bytes.NewReader(data))
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse ELF file: %v", err)
+	}
+
+	section := elfFile.Section(".modinfo")
+	if section == nil {
+		return nil, fmt.Errorf("no .modinfo section found")
+	}
+
+	sectionData, err := section.Data()
+	if err != nil {
+		return nil, fmt.Errorf("failed to read .modinfo section: %v", err)
+	}
+
+	var (
+		moduleName    string
+		moduleVersion string
+		srcVersion    string
+		verMagic      string
+		author        string
+		depends       []string
+	)
+
+	for _, line := range bytes.Split(sectionData, []byte{'\x00'}) {
+		if entry := strings.SplitN(string(line), "=", 2); len(entry) == 2 {
```

if the length isn't 2 is that a malformed binary? Should we return an error in that case?

---

The length should always be 2. I added the check since in the following line I access key and value with entry[0] and entry[1].

`entry` should always be in the form key=value (e.g. `description=Power-button driver for Basin Cove PMIC`)

---

There’s an edge case where the last line of the .modinfo section is always empty. Throwing an error for len != 2 would cause us to skip the parsing of the module information entirely.

Would it be ok to first check if the line is empty and then check if the length is != 2?
```
if len(line) == 0 {
    continue // Skip empty lines
}

if len(entry) != 2 {
    return nil, fmt.Errorf("malformed .modinfo entry, expected 'key=value' but got: %s", string(line))
}
```